crates/bevy_ecs/src/change_detection.rs

```diff
@@ -75,6 +82,24 @@ pub trait DetectChanges {
     /// [`SystemParam`](crate::system::SystemParam).
     fn last_changed(&self) -> u32;
 
+    /// Sets `self` to `value`, if and only if `*self != *value`
+    ///
+    /// `T` is the type stored within the smart pointer (e.g. [`Mut`] or [`ResMut`]).
+    ///
+    /// This is useful to ensure change detection is only triggered when the underlying value
+    /// changes, instead of every time [`DerefMut`] is used.
+    #[inline]
+    fn set_if_neq<T>(&mut self, value: T)
+    where
+        Self: Deref<Target = T> + DerefMut<Target = T>,
+        T: PartialEq,
+    {
+        // This dereference is immutable, so does not trigger change detection
+        if **self != value {
+            **self = value;
+        }
```

Comment:
Perhaps explicit `<Self as Deref<Target = T>>::deref(*self, ...)` and `<Self as DerefMut<Target = T>>::deref_mut(*self, ...)`  would be beneficial for clarity?

This pattern may be common, but it's rather opaque if you aren't aware that Rust will not cache one DerefMut  implicitly to share across both references to `**self`.
It's also a bit clearer than trying to explain that one is "immutable" as a prior commenter mentioned, as it becomes explicit; then you can just comment "triggers change tracking" or similar on the second one.

Reply:
Just to weigh in, I think it's fine as-is.